src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -322,6 +324,299 @@ public virtual void Encode(TextWriter output, char[] value, int startIndex, int
+
+                        fixed (char* pDestination = &MemoryMarshal.GetReference(destination))
+                        {
+                            if (TryEncodeUnicodeScalar((int)nextScalarValue, pDestination, destination.Length, out int charsWrittenJustNow))
+                            {
+                                destination = destination.Slice(charsWrittenJustNow); // advance destination buffer
+                            }
+                            else
+                            {
+                                goto ReturnDestinationTooSmall; // Not enough room in the destination buffer to write the transcoded output.
+                            }
+                        }
+
+                        break;
+                }
+
+                source = source.Slice(charsConsumedThisIteration);
+            }
+
+            // Input buffer has been fully processed!
+
+            charsConsumed = originalSourceLength;
+            charsWritten = originalDestinationLength - destination.Length;
+            return OperationStatus.Done;
+
+        ReturnDestinationTooSmall:
+
+            charsConsumed = originalSourceLength - source.Length;
+            charsWritten = originalDestinationLength - destination.Length;
+            return OperationStatus.DestinationTooSmall;
+
+        ReturnNeedMoreData:
+
+            charsConsumed = originalSourceLength - source.Length;
+            charsWritten = originalDestinationLength - destination.Length;
+            return OperationStatus.NeedMoreData;
+        }
+
+        public unsafe virtual OperationStatus EncodeUtf8(ReadOnlySpan<byte> utf8Source, Span<byte> utf8Destination, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true)
+        {
+            // Optimization: Detect how much "doesn't require escaping" data exists at the beginning of the buffer,
+            // and memcpy it directly to the destination.
+
+            int numBytesToCopy = FindFirstCharacterToEncodeUtf8(utf8Source);
+            if (numBytesToCopy < 0)
+            {
+                numBytesToCopy = utf8Source.Length;
+            }
+
+            if (!utf8Source.Slice(0, numBytesToCopy).TryCopyTo(utf8Destination))
+            {
+                // There wasn't enough room in the destination to copy over the entire source buffer.
+                // We'll instead copy over as much as we can and return Incomplete. We do need to
+                // account for the fact that we don't want to truncate a multi-byte UTF-8 subsequence
+                // mid-sequence (since a subsequent slice and call to EncodeUtf8 would produce invalid
+                // data).
+
+                utf8Source = utf8Source.Slice(0, utf8Destination.Length + 1); // guaranteed not to fail since utf8Source is larger than utf8Destination
+                for (int i = utf8Source.Length - 1; i >= 0; i--)
+                {
+                    if (!UnicodeHelpers.IsUtf8ContinuationByte(in utf8Source[i]))
```

Comment:
~Why pass this as `in`?~

Nevermind, saw the comment in `IsUtf8ContinuationByte`)